packages/pubsub/src/subscription.js

```diff
@@ -592,150 +395,315 @@ Subscription.prototype.decorateMessage_ = function(message) {
+ */
+Subscription.prototype.modifyPushConfig = function(config, gaxOpts, callback) {
+  if (is.fn(gaxOpts)) {
+    callback = gaxOpts;
+    gaxOpts = {};
+  }
 
   var reqOpts = {
     subscription: this.name,
-    returnImmediately: !!options.returnImmediately,
-    maxMessages: options.maxResults
+    pushConfig: config
   };
 
-  this.activeRequest_ = this.parent.request(protoOpts, reqOpts, function(err) {
-    self.activeRequest_ = null;
+  this.request({
+    client: 'subscriberClient',
+    method: 'modifyPushConfig',
+    reqOpts: reqOpts,
+    gaxOpts: gaxOpts
+  }, callback);
+};
 
-    var resp = arguments[1];
+/**
+ *
+ */
+Subscription.prototype.nack_ = function(message) {
+  this.breakLease_(message);
+
+  if (!this.connectionPool) {
+    this.inventory_.nack.push(message.ackId);
+    this.setFlushTimeout_();
+    return;
+  }
 
+  this.connectionPool.acquire(message.connectionId, function(err, connection) {
     if (err) {
-      if (err.code === 504) {
-        // Simulate a server timeout where no messages were received.
-        resp = {
-          receivedMessages: []
-        };
-      } else {
-        callback(err, null, resp);
-        return;
-      }
+      self.emit('error', err);
+      return;
     }
+console.log('nacking');
+    connection.write({
+      modifyDeadlineAckIds: [message.ackId],
+      modifyDeadlineSeconds: [0]
+    });
+  });
+}
 
-    var messages = arrify(resp.receivedMessages)
-      .map(function(msg) {
-        return Subscription.formatMessage_(msg, self.encoding);
-      })
-      .map(self.decorateMessage_.bind(self));
+/**
+ * @private
+ */
+Subscription.prototype.openConnection_ = function() {
+  var self = this;
 
-    self.refreshPausedStatus_();
+  var pool = this.connectionPool = new ConnectionPool(this, {
+    ackDeadline: this.ackDeadline,
+    maxConnections: this.maxConnections
+  });
 
-    if (self.autoAck && messages.length !== 0) {
-      var ackIds = messages.map(prop('ackId'));
+  pool.on('error', function(err) {
+    self.emit('error', err);
+  });
+
+  pool.on('message', function(message) {
+    self.emit('message', self.leaseMessage_(message));
+
+    if (self.hasMaxMessages_()) {
+      pool.pause();
+    }
+  });
 
-      self.ack(ackIds, function(err) {
-        callback(err, messages, resp);
+  pool.once('connected', function() {
+    clearTimeout(self.flushTimeoutHandle_);
+    self.flushTimeoutHandle_ = null;
+    self.flushQueues_();
+  });
+};
+
+/**
+ *
+ */
+Subscription.prototype.renewLeases_ = function() {
+  var self = this;
+
+  this.leaseTimeoutHandle_ = null;
+
+  if (!this.inventory_.lease.length) {
+    return;
+  }
+
+  var ackIds = this.inventory_.lease;
+  this.ackDeadline = this.histogram.percentile(99);
+  var ackDeadlineSeconds = this.ackDeadline / 1000;
+
+  if (this.connectionPool) {
+    this.connectionPool.acquire(function(err, connection) {
+      if (err) {
+        self.emit('error', err);
+        return;
+      }
+console.log({
```

Comment:
No, this was some debug code I accidentally committed! I haven't linted the code yet, usually I do that once I start writing tests, but we do not ship code containing console logs.